src/main/java/org/truffleruby/core/GCNodes.java

```diff
@@ -122,15 +121,100 @@ public static int getCollectionCount() {
+        @Specialization
+        protected RubyArray stat() {
             long time = 0;
+            int count = 0;
+            int minorCount = 0;
+            int majorCount = 0;
+            int unknownCount = 0;
+            String[] memoryPoolNames = new String[0];
+            Object[] memoryPools;
+
+            // Get GC time and counts from GarbageCollectorMXBean
             for (GarbageCollectorMXBean bean : ManagementFactory.getGarbageCollectorMXBeans()) {
+                // Get MemoryPoolName relevant to GC
+                if (bean.getMemoryPoolNames().length > memoryPoolNames.length) {
+                    // Since old generation memory pools are a superset of young generation memory pools,
+                    // it suffices to check that we have the longer list of memory pools
+                    memoryPoolNames = bean.getMemoryPoolNames();
+                }
                 time += bean.getCollectionTime();
+                count += bean.getCollectionCount();
+                switch (bean.getName()) {
+                    case "G1 Young Generation": // during 'jvm' and 'jvm-ce'
+                    case "PS Scavenge": // during 'jvm --vm.XX:+UseParallelGC'
+                    case "young generation scavenger": // during 'native'
+                        minorCount += bean.getCollectionCount();
+                        break;
+                    case "G1 Old Generation": // during 'jvm' and 'jvm-ce'
+                    case "PS MarkSweep": // during 'jvm --vm.XX:+UseParallelGC'
+                    case "complete scavenger": // during 'native'
+                        majorCount += bean.getCollectionCount();
+                        break;
+                    default:
+                        unknownCount += bean.getCollectionCount();
+                        break;
+                }
             }
-            return time;
+
+            // Get memory usage values from relevant memory pools (2-3 / ~8 are relevant)
+            memoryPools = new Object[memoryPoolNames.length];
+            for (MemoryPoolMXBean bean : ManagementFactory.getMemoryPoolMXBeans()) {
+                if (bean.getName().equals(memoryPoolNames[0])) {
+                    memoryPools[0] = beanToArray(bean);
+                } else if (bean.getName().equals(memoryPoolNames[1])) {
+                    memoryPools[1] = beanToArray(bean);
+                } else if (memoryPoolNames.length == 3 && bean.getName().equals(memoryPoolNames[2])) {
+                    memoryPools[2] = beanToArray(bean);
+                }
+            }
+
+            MemoryUsage usage = ManagementFactory.getMemoryMXBean().getHeapMemoryUsage();
+
+            // Use an object array instead because otherwise ArrayHelpers.java line 51 complains
+            Object[] memoryPoolNamesCast = new Object[memoryPoolNames.length];
+            for (int i = 0; i < memoryPoolNames.length; i++) {
+                memoryPoolNamesCast[i] = memoryPoolNames[i];
+            }
```

Comment:
This code seems somewhat redundant, but casting String[] to Object[] was ineffective for me, and there are various assertion where the store of a RubyArray is Object[] can cannot be a subclass of Object[].

Reply:
Yes, this is necessary, otherwise one couldn't write e.g. an Integer to the Ruby Array.
There is still an issue here that Ruby String should be stored in the array, not Java Strings.
Use MakeStringNode for that, like `makeStringNode.executeMake(memoryPoolNames[i], UTF8Encoding.INSTANCE, CodeRange.CR_UNKNOWN)`

Reply:
Thanks!! I originally looked for how to do that but was unsuccessful.